crates/astria-sequencer/src/service/mempool/mod.rs

```diff
@@ -211,179 +134,179 @@ impl Service<MempoolRequest> for Mempool {
 
-    let tx_hash = sha2::Sha256::digest(&tx).into();
-
-    // check if the transaction has been removed from the appside mempool
-    if let Err(rsp) = check_removed_comet_bft(tx_hash, mempool, metrics).await {
-        return rsp;
-    }
-
-    // check if the transaction is already in the mempool
-    if is_tracked(tx_hash, mempool, metrics).await {
-        return response::CheckTx::default();
-    }
-
-    // perform stateless checks
-    let signed_tx = match stateless_checks(tx, &state, metrics).await {
-        Ok(signed_tx) => signed_tx,
-        Err(rsp) => return rsp,
+    let signed_tx = match convert_bytes_to_signed_tx(tx, metrics).await {
+        Ok(tx) => tx,
+        Err(err) => return err.into_check_tx_response(),
     };
 
-    // attempt to insert the transaction into the mempool
-    if let Err(rsp) = insert_into_mempool(mempool, &state, signed_tx, metrics).await {
-        return rsp;
-    }
-
-    // insertion successful
-    metrics.set_transactions_in_mempool_total(mempool.len().await);
-
-    response::CheckTx::default()
-}
-
-/// Checks if the transaction is already in the mempool.
-#[instrument(skip_all, fields(tx_hash = %hex::encode(tx_hash)))]
-async fn is_tracked(tx_hash: [u8; 32], mempool: &AppMempool, metrics: &Metrics) -> bool {
-    let start_tracked_check = Instant::now();
-
-    let result = mempool.is_tracked(tx_hash).await;
-
-    metrics.record_check_tx_duration_seconds_check_tracked(start_tracked_check.elapsed());
-
-    result
-}
-
-/// Checks if the transaction has been removed from the appside mempool.
-///
-/// Returns an `Err(response::CheckTx)` with an error code and message if the transaction has been
-/// removed from the appside mempool.
-#[instrument(skip_all, fields(tx_hash = %hex::encode(tx_hash)))]
-async fn check_removed_comet_bft(
-    tx_hash: [u8; 32],
-    mempool: &AppMempool,
-    metrics: &Metrics,
-) -> Result<(), response::CheckTx> {
-    let start_removal_check = Instant::now();
-
-    // check if the transaction has been removed from the appside mempool and handle
-    // the removal reason
-    if let Some(removal_reason) = mempool.check_removed_comet_bft(tx_hash).await {
-        match removal_reason {
+    let outcome = check_tx(signed_tx.clone(), state, mempool, metrics).await;
+    if let CheckTxOutcome::RemovedFromMempool(reason) = outcome {
+        match reason {
             RemovalReason::Expired => {
                 metrics.increment_check_tx_removed_expired();
-                return Err(removal_reason.into_check_tx_response());
             }
             RemovalReason::FailedPrepareProposal(_) => {
                 metrics.increment_check_tx_removed_failed_execution();
-                return Err(removal_reason.into_check_tx_response());
             }
-            _ => return Err(removal_reason.into_check_tx_response()),
+            _ => {}
+        };
+        mempool
+            .remove_from_removal_cache(signed_tx.id().get())
+            .await;
```

Comment:
This should also be called for `CheckTxOutcome::IncludedInBlock`, but only noting here in case we decide to not remove that variant (as per suggestion in other comment) for some reason.  If we remove `CheckTxOutcome::IncludedInBlock`, then this comment can be ignored.